### PR TITLE
Added test TEST-40174, TEST-40042

### PR DESCRIPTION
**Pull request by priyankaborawake**
Signed-off-by: priyanka.borawake <priyanka.borawake@seagate.com>

# Problem Statement
-Automated IO stability test TEST-40174, TEST-40042

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [ ] Attach test execution logs
-  [ ] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Review by kulrk — APPROVED** [review]

File: tests/iostability/test_iostability_degraded_path.py [hunk 1]
```diff
@@ -116,3 +124,143 @@ def test_object_crud_single_pod_failure(self):
...
+                      "and perform Object CRUD(40% write,Read,20% delete) operation(degraded mode)")
+
+        duration_in_days = self.test_cfg['degraded_path_durations_days']
+        clients = (len(self.worker_node_list)-1) * self.test_cfg['sessions_per_node_vm']
+        write_percent_per_iter = self.test_cfg['test_40174']['write_percent_per_iter']
+        delete_percent_per_iter = self.test_cfg['test_40174']['delete_percent_per_iter']
+        max_cluster_capacity_percent = self.test_cfg['test_40174']['max_cluster_capacity_percent']
+        s3userinfo = dict()
+        s3userinfo['accesskey'] = ACCESS_KEY
+        s3userinfo['secretkey'] = SECRET_KEY
```
Comment (ketanarlulkar) [inline tests/iostability/test_iostability_degraded_path.py:144]: Prefer a dictionary creation as dictionary literal
```suggestion
        s3userinfo = {'accesskey': ACCESS_KEY, 'secretkey': SECRET_KEY}
```

File: tests/iostability/test_iostability_degraded_path.py [hunk 2]
```diff
@@ -116,3 +124,143 @@ def test_object_crud_single_pod_failure(self):
...
+        bucket_prefix = "test-40174-bkt"
+        end_time = datetime.now() + timedelta(days=duration_in_days)
+
+        self.log.info("Step 1: Get the used percentage of disk space.")
+        total_cap, avail_cap, used_cap = self.health_obj.get_sys_capacity()
```
Comment (ketanarlulkar) [inline tests/iostability/test_iostability_degraded_path.py:149]: Does `get_sys_capacity()` give values by taking SNS configuration in account?

Comment (priyankaborawake) [reply tests/iostability/test_iostability_degraded_path.py]: No, get_sys_capacity does not take SNS configuration in account.

File: tests/iostability/test_iostability_degraded_path.py [hunk 3]
```diff
@@ -220,3 +219,169 @@ def test_disk_near_full_read_in_degraded_s3bench(self):
         self.test_completed = True
         self.log.info("ENDED: Perform disk storage near full once and read in loop for %s days",
                       self.duration_in_days)
+
+    @pytest.mark.lc
+    @pytest.mark.io_stability
+    @pytest.mark.tags("TEST-40174")
+    def test_degraded_iteration_write_read_partial_delete(self):
```
Comment (seagate-sarang-sawant) [inline tests/iostability/test_iostability_degraded_path.py:226]: This test is very long. Can you divide into more reusable functions and use them in similar tests in this module.

File: tests/iostability/test_iostability_happy_path.py [hunk 4]
```diff
@@ -195,3 +196,114 @@ def test_disk_near_full_s3bench(self):
         self.test_completed = True
         self.log.info("ENDED: Perform disk storage near full once and read in loop for %s days",
                       self.duration_in_days)
+
+    @pytest.mark.lc
+    @pytest.mark.io_stability
+    @pytest.mark.tags("TEST-40042")
+    def test_iteration_write_read_partial_delete(self):
+        """Perform iterations of 30% writes of user capacity ,
```
Comment (seagate-sarang-sawant) [inline tests/iostability/test_iostability_happy_path.py:204]: Can you divide into more reusable functions and use them in similar tests in this module?

**Review by seagate-sarang-sawant — CHANGES_REQUESTED** [review]

File: tests/iostability/test_iostability_degraded_path.py (hunk 3 above)
Comment (priyankaborawake) [reply tests/iostability/test_iostability_degraded_path.py:225]: Added two function in lib.
perform_write_to_fill_system_percent and delete_workload

File: tests/iostability/test_iostability_happy_path.py (hunk 4 above)
Comment (priyankaborawake) [reply tests/iostability/test_iostability_happy_path.py:203]: Added two function in lib.
perform_write_to_fill_system_percent and delete_workload

**Review by ketanarlulkar — APPROVED** [review]

File: libs/durability/near_full_data_storage.py [hunk 5]
```diff
@@ -82,25 +84,33 @@ def perform_near_full_sys_writes(s3userinfo, user_data_writes, bucket_prefix: st
...
+        :param client (Optional): Number of client sessions
+        :param bucket_list (Optional): List of pre-created buckets.
+        :return : boolean, list of dictionary
                 format : [{'bucket': bucket_name, 'obj_name_pref': obj_name, 'num_clients':
                 client, 'obj_size': obj_size, 'num_sample': sample}]
         """
         client = kwargs.get("client", 10)
-        workload = [128, 256, 512]  # workload in mb
+        bucket_list = kwargs.get('bucket_list', None)
+        workload = [128, 256, 512, 1024, 2048]  # workload in mb
```
[the remark this answers is not in the document]
Comment (priyankaborawake) [reply libs/durability/near_full_data_storage.py:95]: Sure will move it to config..

File: tests/iostability/test_iostability_degraded_path.py [hunk 6]
```diff
@@ -220,3 +218,118 @@ def test_disk_near_full_read_in_degraded_s3bench(self):
...
+                        s3userinfo=s3userinfo, workload_info=workload_info_list,
+                        skipread=False, validate=True, skipcleanup=True)
+                    assert_utils.assert_true(resp[0], resp[1])
+                else:
+                    self.log.warning("No buckets available to perform read operations %s",
+                                     workload_info_list)
+
+                self.log.info("Delete %s percent of the written data")
+                if len(workload_info_list) > 0:
+                    resp = NearFullStorage.delete_workload(workload_info_list, s3userinfo,
```
[the remark this answers is not in the document]
Comment (priyankaborawake) [reply tests/iostability/test_iostability_degraded_path.py:313]: @sonalk0209 in 302-309 we are reading all the available data in all the buckets , in delete method, we are deleting only few buckets.

File: tests/iostability/test_iostability_degraded_path.py [hunk 7]
```diff
@@ -220,3 +218,118 @@ def test_disk_near_full_read_in_degraded_s3bench(self):
...
+
+        self.mail_notify = send_mail_notification(self.sender_mail_id, self.receiver_mail_id,
+                                                  cortxlogging.get_frame(), self.health_obj_list[0])
+
+        max_percentage = self.test_cfg['nearfull_storage_percentage']
+        clients = (len(self.worker_node_list) - 1) * self.clients
+        s3userinfo = {'accesskey': ACCESS_KEY, 'secretkey': SECRET_KEY}
+        end_time = datetime.now() + timedelta(days=self.duration_in_days)
+
+        self.log.info("Step 1: Create 50 buckets in healthy mode ")
```
[the remark this answers is not in the document]
Comment (priyankaborawake) [reply tests/iostability/test_iostability_degraded_path.py:247]: Moved to iostability config.

File: config/durability/near_full_data_storage.yaml [hunk 8]
```diff
@@ -0,0 +1,3 @@
+full_sys_writes:
+  vm_workload: [128, 256, 512, 1024, 2048]
+  extended_hw_workload: [3072, 4096]
```
Comment (ApurwaMohite) [inline config/durability/near_full_data_storage.yaml:3]: Please add new line at the end.

File: libs/durability/near_full_data_storage.py [hunk 9]
```diff
@@ -20,11 +20,13 @@
...
 """
 import logging
+import random
+import time
 
 from commons.helpers.health_helper import Health
 from commons.helpers.pods_helper import LogicalNode
 from commons.constants import MB
-from config import CMN_CFG
+from config import CMN_CFG, NEAR_FULL_STORAGE_CFG
```
Comment (ApurwaMohite) [inline libs/durability/near_full_data_storage.py:29]: Single import on single line.

File: libs/durability/near_full_data_storage.py [hunk 10]
```diff
@@ -82,25 +84,33 @@ def perform_near_full_sys_writes(s3userinfo, user_data_writes, bucket_prefix: st
...
         if CMN_CFG["setup_type"] == "HW":
-            workload.extend([1024, 2048, 3072, 4096])
+            workload.extend(NEAR_FULL_STORAGE_CFG['full_sys_writes']['extended_hw_workload'])
 
         workload = [each * MB for each in workload]  # convert to bytes
         each_workload_byte = user_data_writes / len(workload)
         return_list = []
+        if bucket_list:
+            bucket_iter = iter(bucket_list)
+        LOGGER.info("Perform Write operation for object size : %s",workload)
```
Comment (ApurwaMohite) [inline libs/durability/near_full_data_storage.py:104]: space after before 'workload'.

File: libs/durability/near_full_data_storage.py [hunk 11]
```diff
@@ -158,3 +169,71 @@ def perform_near_full_sys_operations(s3userinfo, workload_info: list, skipread:
...
+
+    # pylint: disable=too-many-arguments
+    @staticmethod
+    def perform_write_to_fill_system_percent(master_node: LogicalNode, write_per: int, s3userinfo,
+                                             bucket_prefix, clients, bucket_list=None):
+        """
+        Retrieve user data space to attain specific percent of near full storage.
+        :param master_node: Master node object
+        :param write_per: Percentage of near full storage to be attained.
+        :param s3userinfo: User info for S3 user
```
Comment (ApurwaMohite) [inline libs/durability/near_full_data_storage.py:213]: iam * user

File: config/__init__.py [hunk 12]
```diff
@@ -158,7 +158,7 @@ def build_s3_endpoints() -> dict:
 
 DI_CFG = configmanager.get_config_wrapper(fpath=DI_CONFIG_PATH)
 DATA_PATH_CFG = configmanager.get_config_wrapper(fpath=DATA_PATH_CONFIG_PATH, target=target)
-
+NEAR_FULL_STORAGE_CFG = configmanager.get_config_wrapper(fpath=NEAR_FULL_STORAGE_CFG_PATH)
```
Comment (sonalk0209) [inline config/__init__.py:161]: u need to import this first

File: config/durability/near_full_data_storage.yaml [hunk 13]
```diff
@@ -0,0 +1,3 @@
+full_sys_writes:
```
Comment (sonalk0209) [inline config/durability/near_full_data_storage.yaml:1]: Can we rather add a config file which is common for all durability tests? The name can be general enough like test_durability.yaml maybe
And may not require to create under a new dir , it can be under config itself

File: tests/iostability/test_iostability_happy_path.py [hunk 14]
```diff
@@ -195,3 +195,79 @@ def test_disk_near_full_s3bench(self):
...
+        self.log.info(
+            "STARTED: Test for performing %s percent writes, read written data and deleting "
+            "%s percent of written data.", write_percent_per_iter, delete_percent_per_iter)
+
+        test_case_name = cortxlogging.get_frame()
+        self.mail_notify = send_mail_notification(self.sender_mail_id, self.receiver_mail_id,
+                                                  test_case_name, self.health_obj_list[0])
+
+        max_cluster_capacity_percent = self.test_cfg['nearfull_storage_percentage']
+        s3userinfo = {'accesskey': ACCESS_KEY, 'secretkey': SECRET_KEY}
```
Comment (ApurwaMohite) [inline tests/iostability/test_iostability_happy_path.py:217]: This can be added in setup_class.

File: tests/iostability/test_iostability_degraded_path.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (sonalk0209) [reply tests/iostability/test_iostability_degraded_path.py:247]: u can use the same in logger as well

**Review by sonalk0209 — COMMENTED** [review]
@priyankaborawake few minor comments, plz check

File: libs/durability/near_full_data_storage.py [hunk 15]
```diff
@@ -158,3 +171,71 @@ def perform_near_full_sys_operations(s3userinfo, workload_info: list, skipread:
...
+        LOGGER.info("Deleting buckets : %s", delete_list)
+        resp = NearFullStorage.perform_operations_on_pre_written_data(
+            s3userinfo=s3userinfo,
+            workload_info=delete_list,
+            skipread=False,
+            validate=True,
+            skipcleanup=False)
+
+        if resp[0]:
+            deleted_buckets = [each['bucket'] for each in delete_list]
```
Comment (ApurwaMohite) [inline libs/durability/near_full_data_storage.py:202]: why is this needed?

File: tests/iostability/test_iostability_degraded_path.py [hunk 16]
```diff
@@ -209,14 +207,131 @@ def test_disk_near_full_read_in_degraded_s3bench(self):
...
+            write_percent_per_iter, write_percent_per_iter, delete_percent_per_iter)
+
+        self.mail_notify = send_mail_notification(self.sender_mail_id, self.receiver_mail_id,
+                                                  cortxlogging.get_frame(), self.health_obj_list[0])
+
+        max_percentage = self.test_cfg['nearfull_storage_percentage']
+        clients = (len(self.worker_node_list) - 1) * self.clients
+        end_time = datetime.now() + timedelta(days=self.duration_in_days)
+
+        self.log.info("Step 1: Create %s buckets in healthy mode",
```
Comment (ApurwaMohite) [inline tests/iostability/test_iostability_degraded_path.py:247]: This logger should be inside if.

File: tests/iostability/test_iostability_degraded_path.py [hunk 17]
```diff
@@ -209,14 +207,131 @@ def test_disk_near_full_read_in_degraded_s3bench(self):
...
+                if len(workload_info_list) > 0:
+                    resp = NearFullStorage.perform_operations_on_pre_written_data(
+                        s3userinfo=self.s3userinfo, workload_info=workload_info_list,
+                        skipread=False, validate=True, skipcleanup=True)
+                    assert_utils.assert_true(resp[0], resp[1])
+                else:
+                    self.log.warning("No buckets available to perform read operations %s",
+                                     workload_info_list)
+
+                self.log.info("Delete %s percent of the written data")
```
Comment (ApurwaMohite) [inline tests/iostability/test_iostability_degraded_path.py:313]: string/variable needed for formatting.

File: tests/iostability/test_iostability_degraded_path.py [hunk 18]
```diff
@@ -209,14 +207,131 @@ def test_disk_near_full_read_in_degraded_s3bench(self):
...
+                if len(workload_info_list) > 0:
+                    resp = NearFullStorage.delete_workload(workload_info_list, self.s3userinfo,
+                                                           delete_percent_per_iter)
+                    assert_utils.assert_true(resp[0], resp[1])
+                    for each in resp[1]:
+                        avail_buckets.append(each)
+                else:
+                    self.log.warning("No buckets available to perform delete operations %s",
+                                     workload_info_list)
```
Comment (ApurwaMohite) [inline tests/iostability/test_iostability_degraded_path.py:322]: This can be part of above if-else itself as the condition is same.

File: tests/iostability/test_iostability_happy_path.py [hunk 19]
```diff
@@ -177,21 +177,97 @@ def test_disk_near_full_s3bench(self):
...
+
+                # Delete "delete_percent_per_iter" data of all the written data
+                self.log.info("Delete %s percent of the written data", delete_percent_per_iter)
+                if len(workload_info_list) > 0:
+                    resp = NearFullStorage.delete_workload(workload_info_list, self.s3userinfo,
+                                                           delete_percent_per_iter)
+                    assert_utils.assert_true(resp[0], resp[1])
+                else:
+                    self.log.warning("No buckets available to perform delete operations,"
+                                     "workload info list : %s", workload_info_list)
```
Comment (ApurwaMohite) [inline tests/iostability/test_iostability_happy_path.py:261]: Same as above.

File: libs/durability/near_full_data_storage.py (hunk 15 above)
Comment (priyankaborawake) [reply libs/durability/near_full_data_storage.py:202]: As new bucket creation/deletion is not allowed, we are reusing the buckets post deletion . 
hence retrieving bucket names from workload info for reuse.

**Review by sonalk0209 — APPROVED** [review]

**Review by seagate-sarang-sawant — APPROVED** [review]

**Review by ApurwaMohite — APPROVED** [review]
